cpp/src/arrow/util/tdigest.h

```diff
@@ -0,0 +1,419 @@
+    }
+    for (TDigest* td : tdigests) {
+      td->MergeInput();
+      if (td->tdigest_->size() > 0) {
+        queue.emplace(td->tdigest_->cbegin(), td->tdigest_->cend());
+        total_weight_ += td->total_weight_;
+        min_ = std::min(min_, td->min_);
+        max_ = std::max(max_, td->max_);
+      }
+    }
+
+    merger_.Reset(total_weight_, tdigest_next_);
+
+    CentroidIter current_iter, end_iter;
+    // do k-way merge till one buffer left
+    while (queue.size() > 1) {
+      std::tie(current_iter, end_iter) = queue.top();
+      merger_.Add(*current_iter);
+      queue.pop();
+      if (++current_iter != end_iter) {
+        queue.emplace(current_iter, end_iter);
+      }
+    }
+    // merge last buffer
+    if (!queue.empty()) {
+      std::tie(current_iter, end_iter) = queue.top();
+      while (current_iter != end_iter) {
+        merger_.Add(*current_iter++);
+      }
+    }
+
+    std::swap(tdigest_, tdigest_next_);
+  }
+
+  // an ugly helper
+  void Merge(std::vector<std::unique_ptr<TDigest>>& ptrs) {
+    std::vector<TDigest*> tdigests;
+    tdigests.reserve(ptrs.size());
+    for (auto& ptr : ptrs) {
+      tdigests.push_back(ptr.get());
+    }
+    Merge(tdigests);
+  }
+
+  // calculate quantile
+  double Quantile(double q) {
+    MergeInput();
+
+    const auto& td = *tdigest_;
+
+    if (q < 0 || q > 1) {
+      ARROW_LOG(ERROR) << "quantile must be between 0 and 1";
+      return NAN;
+    }
+    if (td.size() == 0) {
+      ARROW_LOG(WARNING) << "empty tdigest";
+      return NAN;
+    }
+
+    const double index = q * total_weight_;
+    if (index <= 1) {
+      return min_;
+    } else if (index >= total_weight_ - 1) {
+      return max_;
+    }
+
+    // find centroid contains the index
+    uint32_t ci = 0;
+    double weight_sum = 0;
+    for (; ci < td.size(); ++ci) {
+      weight_sum += td[ci].weight;
+      if (index <= weight_sum) {
+        break;
+      }
+    }
+    DCHECK_LT(ci, td.size());
+
+    // deviation of index from the centroid center
+    double diff = index + td[ci].weight / 2 - weight_sum;
+
+    // index happen to be in a unit weight centroid
+    if (td[ci].weight == 1 && std::abs(diff) < 0.5) {
+      return td[ci].mean;
+    }
+
+    // find adjacent centroids for interpolation
+    uint32_t ci_left = ci, ci_right = ci;
+    if (diff > 0) {
+      if (ci_right == td.size() - 1) {
+        // index larger than center of last bin
+        DCHECK_EQ(weight_sum, total_weight_);
+        const detail::Centroid* c = &td[ci_right];
+        DCHECK_GE(c->weight, 2);
+        return detail::Lerp(c->mean, max_, diff / (c->weight / 2));
+      }
+      ++ci_right;
+    } else {
+      if (ci_left == 0) {
+        // index smaller than center of first bin
+        const detail::Centroid* c = &td[0];
+        DCHECK_GE(c->weight, 2);
+        return detail::Lerp(min_, c->mean, index / (c->weight / 2));
+      }
+      --ci_left;
+      diff += td[ci_left].weight / 2 + td[ci_right].weight / 2;
```

Comment:
It seems this would be the same as doing `if (diff < 0) { diff += 1; }` below?
(not sure which one is more readable or more faithful to the underlying math)

Reply:
Good catch!
I think they are both faithful to math. `diff` is finally the normalized distance from quantile point to left centroid (taken distance from left to right centroids as 1). Current code calculates it by the definition. Your suggestion calculates it as `1 - distance-to-right-centroid`.
Suggested code is more concise, drop that somewhat duplicated line, add one simple line `diff += (diff <= 0)`).
Current code is more straightforward for me. I would prefer keeping current code. And I guess compiler can do `common subexpression elimination` to avoid duplicated calculation?